File: Python/OneLang/VM/Values.py
from onelang_core import *

class ObjectValue:
    def __init__(self, props):
        self.props = props
    
    def equals(self, other):
        return False

class StringValue:
    def __init__(self, value):
        self.value = value
    
    def equals(self, other):
        return isinstance(other, StringValue) and other.value == self.value

class NumericValue:
    def __init__(self, value):
        self.value = value
    
    def equals(self, other):
        return isinstance(other, NumericValue) and other.value == self.value

class BooleanValue:
    def __init__(self, value):
        self.value = value
    
    def equals(self, other):
        return isinstance(other, BooleanValue) and other.value == self.value

class ArrayValue:
    def __init__(self, items):
        self.items = items
    
    def equals(self, other):
        return False

File: PHP/src/OneLang/VM/Values.php
<?php

namespace OneLang\VM\Values;

interface IVMValue {
    function equals($other);
}

interface ICallableValue extends IVMValue {
    function call($args);
}

class ObjectValue implements IVMValue {
    public $props;
    
    function __construct($props) {
        $this->props = $props;
    }
    
    function equals($other) {
        return false;
    }
}

class StringValue implements IVMValue {
    public $value;
    
    function __construct($value) {
        $this->value = $value;
    }
    
    function equals($other) {
        return $other instanceof StringValue && $other->value === $this->value;
    }
}

class NumericValue implements IVMValue {
    public $value;
    
    function __construct($value) {
        $this->value = $value;
    }
    
    function equals($other) {
        return $other instanceof NumericValue && $other->value === $this->value;
    }
}

class BooleanValue implements IVMValue {
    public $value;
    
    function __construct($value) {
        $this->value = $value;
    }
    
    function equals($other) {
        return $other instanceof BooleanValue && $other->value === $this->value;
    }
}

class ArrayValue implements IVMValue {
    public $items;
    
    function __construct($items) {
        $this->items = $items;
    }
    
    function equals($other) {
        return false;
    }
}
